src/utils/analytics.utils.ts

```diff
@@ -0,0 +1,117 @@
+import { FilterDto } from 'src/analytics/dto/filter.dto';
+
+export const buildQueryBuilder = (
+  entityManager: EntityManager,
+  userId: string,
+  filter: FilterDto,
+) => {
+  const queryBuilder = entityManager
+    .getRepository(User)
+    .createQueryBuilder('user')
+    .leftJoinAndSelect('user.repo', 'repo')
+    .leftJoinAndSelect('repo.pullRequests', 'pullRequest')
+    .where('user.githubUserId = :githubUserId', { githubUserId: userId });
+
+  if (filter?.repoName) {
+    queryBuilder.andWhere('repo.name = :name', { name: filter.repoName });
+  }
+
+  return queryBuilder;
+};
+
+export const setDateFilters = (filter: FilterDto, frequency: Frequency) => {
+  let startDate: string;
+  let endDate: string;
+  const now = new Date();
+
+  if (filter?.startDate && filter?.endDate) {
+    const start = new Date(filter.startDate);
+    const end = new Date(filter.endDate);
+    start.setHours(0, 0, 0, 0);
+    end.setHours(23, 59, 59, 999);
+    startDate = start.toISOString();
+    endDate = end.toISOString();
+  } else {
+    let start;
+    if (frequency === Frequency.MONTH) {
+      start = addMonths(now, -2);
+    } else if (frequency === Frequency.YEAR) {
+      start = startOfYear(subYears(now, 2));
+    } else {
+      start = addDays(now, -6);
+    }
+    start.setHours(0, 0, 0, 0);
+    now.setHours(23, 59, 59, 999);
+    startDate = start.toISOString();
+    endDate = now.toISOString();
+  }
+
+  return { startDate, endDate };
+};
+
+export const setGroupBy = (frequency: Frequency) => {
+  if (frequency === Frequency.MONTH) {
+    return "TO_CHAR(event.createdAt, 'YYYY-MM')";
+  } else if (frequency === Frequency.YEAR) {
+    return "TO_CHAR(event.createdAt, 'YYYY')";
+  }
+  return "TO_CHAR(event.createdAt, 'YYYY-MM-DD')";
+};
+
+export const formatData = (
+  data: { period: string; value: number }[],
+  startDate: string,
+  endDate: string,
+  frequency: Frequency,
+) => {
+  const periods = generatePeriods(
+    new Date(startDate),
+    new Date(endDate),
+    frequency,
+  );
+
+  const dataMap = new Map(data.map((entry) => [entry.period, entry.value]));
+
+  return periods.map((period) => ({
+    label:
+      frequency === Frequency.MONTH
+        ? Months[new Date(period).getMonth()]
+        : period,
+    data: dataMap.get(period) || 0,
+  }));
+};
+
+export function generatePeriods(
+  startDate: Date,
+  endDate: Date,
+  frequency: Frequency,
+): string[] {
+  const periods = [];
+  let currentDate = startDate;
+
+  while (currentDate <= endDate) {
+    if (frequency === Frequency.MONTH) {
+      periods.push(format(currentDate, 'yyyy-MM'));
+      currentDate = addMonths(currentDate, 1);
+    } else if (frequency === Frequency.YEAR) {
+      periods.push(format(currentDate, 'yyyy'));
+      currentDate = addYears(currentDate, 1);
+    } else {
+      periods.push(format(currentDate, 'yyyy-MM-dd'));
+      currentDate = addDays(currentDate, 1);
+    }
+  }
+
+  return periods;
```

Comment:
PrimathonAICodeReviewer

The `generatePeriods` function generates periods based on the start date, end date, and frequency provided. The function iterates over dates correctly based on the frequency specified.

#### Overall Comments
The functions in `analytics.utils.ts` are well-segmented based on their responsibilities, making the codebase modular and easier to maintain. The logic within each function is clear and follows best practices for readability and maintainability.

The code adheres to the single responsibility principle, ensuring that each function has a clear purpose and does not take on unnecessary tasks. This approach enhances the overall quality of the codebase and makes it easier to extend or modify in the future.

The functions demonstrate good use of TypeScript type annotations, enhancing code readability and providing type safety during development.

Overall, the changes in `analytics.utils.ts` appear well-structured and maintainable, with a focus on clear and concise functions for specific tasks.


<!-- This is an auto-generated comment by OSS PrimathonAICodeReviewer -->